test/e2e/persistent_volumes.go

```diff
@@ -738,6 +739,109 @@ var _ = framework.KubeDescribe("PersistentVolumes", func() {
+
+		var (
+			volumePath string
+			pv        *v1.PersistentVolume
+			pvc       *v1.PersistentVolumeClaim
+			clientPod *v1.Pod
+			pvConfig  persistentVolumeConfig
+		)
+		BeforeEach(func() {
+			framework.SkipUnlessProviderIs("vsphere")
+			if volumePath == "" {
+				vsp, err := vsphere.GetVSphere()
+				Expect(err).NotTo(HaveOccurred())
+				var volumeoptions vsphere.VolumeOptions
+				volumeoptions.CapacityKB=1 * 1024 * 1024
+				volumeoptions.Name="e2e-disk" + time.Now().Format("20060102150405")
+				volumeoptions.DiskFormat="thin"
+
+				volumePath, err = vsp.CreateVolume(&volumeoptions)
+				Expect(err).NotTo(HaveOccurred())
+				pvConfig = persistentVolumeConfig{
+					namePrefix: "vspherepv-",
+					pvSource: v1.PersistentVolumeSource{
+						VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+							VolumePath:	volumePath,
+							FSType:		"ext4",
+						},
+					},
+					prebind: nil,
+				}
+			}
+		})
+
+		AfterEach(func() {
+			framework.Logf("AfterEach: Cleaning up test resources")
+			if c != nil {
+				deletePod(f, c, ns, clientPod)
+				pvPvcCleanup(c, ns, pv, pvc)
+				clientPod = nil
+				pvc = nil
+				pv = nil
+			}
+		})
+
+		AddCleanupAction(func() {
+			if len(volumePath) > 0 {
+				vsp, err := vsphere.GetVSphere()
+				Expect(err).NotTo(HaveOccurred())
+				vsp.DeleteVolume(volumePath)
+			}
+		})
+
+		// Attach a persistent disk to a pod using a PVC.
+		// Delete the PVC and then the pod.  Expect the pod to succeed in unmounting and detaching PD on delete.
+		It("should test that deleting a PVC before the pod does not cause pod deletion to fail on PD detach", func() {
+			By("Creating the PV and PVC")
+			pv, pvc = createPVPVC(c, pvConfig, ns, false)
+			waitOnPVandPVC(c, ns, pv, pvc)
+
+			By("Creating the Client Pod")
+			clientPod = createClientPod(c, ns, pvc)
+			node := types.NodeName(clientPod.Spec.NodeName)
+
+			By("Deleting the Claim")
+			deletePersistentVolumeClaim(c, pvc.Name, ns)
+			verifyVsphereDiskAttached(volumePath, node)
+
+			By("Deleting the Pod")
+			deletePod(f, c, ns, clientPod)
+
+			By("Verifying Persistent Disk detach")
+			waitForVSphereDiskToDetach(volumePath, node)
+		})
+
+		// Attach a persistent disk to a pod using a PVC.
+		// Delete the PV and then the pod.  Expect the pod to succeed in unmounting and detaching PD on delete.
+		It("should test that deleting the PV before the pod does not cause pod deletion to fail on PD detach", func() {
+			By("Creating the PV and PVC")
+			pv, pvc = createPVPVC(c, pvConfig, ns, false)
+			waitOnPVandPVC(c, ns, pv, pvc)
+
+			By("Creating the Client Pod")
+			clientPod = createClientPod(c, ns, pvc)
+			node := types.NodeName(clientPod.Spec.NodeName)
+
+			By("Deleting the Persistent Volume")
+			deletePersistentVolume(c, pv.Name)
+			verifyVsphereDiskAttached(volumePath, node)
+
+			By("Deleting the client pod")
+			deletePod(f, c, ns, clientPod)
+
+			By("Verifying Persistent Disk detaches")
+			waitForVSphereDiskToDetach(volumePath, node)
```

Comment:
Could not find any such method in vsphere_volume_util.go

Found  similar method in vsphere_volume/attacher.go 
https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/vsphere_volume/attacher.go#L120
but this method just checks volumepath exists or not. Does not check if vmdk is attached to the node or not.

For this test we need a method to wait until disk is detached from node or worker vm. We need to wait until vmdk device is detached from the worker vm, before calling delete volume from CleanupAction.